projects/stone_cx/robot.cc

```diff
@@ -1,170 +1,413 @@
-using namespace BoBRobotics::ImgProc;
-using namespace BoBRobotics::Robots;
-using namespace BoBRobotics::Video;
 
 //---------------------------------------------------------------------------
 // Anonymous namespace
 //---------------------------------------------------------------------------
 namespace
 {
-void buildOpticalFlowFilter(cv::Mat &filter, float preferredAngle) {
-    // Loop through columns
-    for(int x = 0; x < filter.cols; x++) {
-        // Convert column to angle
-        const float th = (((float)x / (float)filter.cols) * 2.0f * Parameters::pi) - Parameters::pi;
-
-        // Write filter with sin of angle
-        filter.at<float>(0, x) = sin(th - preferredAngle);
-    }
-}
 
-void imuThreadFunc(std::atomic<bool> &shouldQuit, std::atomic<float> &heading, unsigned int &numSamples)
+//---------------------------------------------------------------------------
+// HeadingSource
+//---------------------------------------------------------------------------
+//! Interface for ways of getting heading
+class HeadingSource
 {
-    // Create IMU interface
-    LM9DS1 imu;
+public:
+    virtual radian_t getHeading() = 0;
+};
 
-    // Initialise IMU magnetometer
-    LM9DS1::MagnetoSettings magSettings;
-    imu.initMagneto(magSettings);
+//---------------------------------------------------------------------------
+// SpeedSource
+//---------------------------------------------------------------------------
+//! Interface for ways of getting velocity
+class SpeedSource
+{
+public:
+    virtual std::array<float, 2> getSpeed() = 0; 
+};
 
-    // While quit signal isn't set
-    for(numSamples = 0; !shouldQuit; numSamples++) {
+//---------------------------------------------------------------------------
+// MotorController
+//---------------------------------------------------------------------------
+//! Interface to drive robot from model output
+class MotorController
+{
+public:
+    virtual void drive(float left, float right, bool log = false);
+};
+
+//---------------------------------------------------------------------------
+// HeadingSourceLM9DS1IMU
+//---------------------------------------------------------------------------
+class HeadingSourceLM9DS1IMU : public HeadingSource
+{
+public:
+    HeadingSourceLM9DS1IMU()
+    {
+        // Initialise IMU magnetometer was default settings
+        LM9DS1::MagnetoSettings magSettings;
+        m_IMU.initMagneto(magSettings);
+    }
+
+    virtual radian_t getHeading() override
+    {
         // Wait for magneto to become available
-        while(!imu.isMagnetoAvailable()){
+        while(!m_IMU.isMagnetoAvailable()){
         }
 
         // Read magneto
         float magnetoData[3];
-        imu.readMagneto(magnetoData);
+        m_IMU.readMagneto(magnetoData);
 
         // Calculate heading angle from magneto data and set atomic value
-        heading = atan2(magnetoData[0], magnetoData[2]);
+        return radian_t(std::atan2(magnetoData[0], magnetoData[2]));
```

Comment:
Or you could use ``units::math::atan2`` here. Either way's fine though

Reply:
I get a ``error: static assertion failed: The quantity y/x must yield a dimensionless ratio.`` if I switch this to ``units::math::atan2`` - any thoughts?